docs/src/main/sphinx/develop/supporting-merge.rst

```diff
@@ -0,0 +1,379 @@
+in the ``MATCHED`` cases, and inserts in the ``NOT MATCHED`` cases.
+
+For some ``RowChangeParadigms``, ``UPDATE`` operations may have been
+translated into the corresponding ``DELETE`` and ``INSERT`` operations
+before ``storeMergedRows(Page)`` was called.
+
+To complete the ``MERGE`` operation, the Trino engine calls
+``ConnectorMetadata.finishMerge(...)``, passing the table
+and that collection of ``Slice`` fragments with information about what
+was changed, and the connector takes appropriate actions, if
+any.
+
+``RowChangeProcessor`` implementation for ``MERGE``
+---------------------------------------------------
+
+In SQL ``MERGE``, each supported ``RowChangeParadigm``
+corresponds to an internal Trino engine class that implements interface
+``RowChangeProcessor``.  ``RowChangeProcessor`` has one interesting method:
+``Page transformPage(Page)``.  The format of the input and output page depend
+on the the SQL operation being implemented.
+
+The connector has no access to the ``RowChangeProcessor`` instance - - it
+is used inside the Trino engine to transform the merge page rows into rows
+to be stored, based on the connector's choice of ``RowChangeParadigm``.
+
+For SQL MERGE, the page supplied to ``transformPage()`` consists of:
+
+* The write redistribution columns if any
+* The ``rowId`` column for the row from the target table if matched, or
+  null if not matched
+* The merge case ``RowBlock``
+* For partitioned or bucketed tables, a hash value column.
+
+The merge case ``RowBlock`` has this layout:
+
+* Blocks for each column in the table, including partition columns, in
+  table column order.
+* A block containing 1 if the target table row was matched by a source
+  row base on the ``MERGE`` ``JOIN`` condition, and 0 if there was not
+  target table row matching the source row using the ``JOIN`` condition.
+  the row starting at 0, or -1 no merge case matched for the row.
+* A block containing the merge case operation, encoded as ``INSERT`` = 1,
+  ``DELETE`` = 2, ``UPDATE`` = 3 and if no merge case matched, -1.
+
+The page returned from ``transformPage`` consists of all table columns,
+in table column order, followed by the rowId column, followed by the
+operation column from the merge case ``RowBlock``.
+
+Interface ``RowChangeProcessor`` now supports SQL MERGE, and will later be
+used to upgrade the SQL UPDATE engine implementation.
+
+Detecting duplicate matching target rows
+----------------------------------------
+
+The SQL ``MERGE`` specification requires that in each ``MERGE`` case,
+a single target table row must match at most one source row, after
+applying the ``MERGE`` case condition expression.  This is done in
+class ``DuplicateRowFinder.checkForDuplicateTargetRows()``.  That method
+examines successive rows produced by the ``SELECT CASE`` expression,
+excluding ``INSERT``ed rows, and if the ``writeRedistributionColumns``
+and the ``rowId`` column are the identical in successive rows.
```

Comment:
@electrum and I convinced ourselves that target table rowIds would be partitioned among nodes, in the sense that the same target rowId would never show up in two different nodes, and that if two target rowIds were identical, they would be adjacent.  I'll consult @electrum to see if our logic was faulty.

Reply:
> i would like to learn about the following
> 
> * what are the assumption on rowIds, can rowIds carry un-updated columns

Yes.  The rowIds from the target table don't get updated - - they represent a durable identity for the target table rows.  Rows written as a result of the merge of course are written with newly-generated rowIds.

> * how should a connector construct rowIds if it needs to create deletion delta files for the sake of updates (e.g. a separate deletion file for an input file which would mark all the rows that got updated)

That's the point of the (poorly named) `DeleteAndInsertMergeProcessor` - - given a page that contains updated rows, it splits each updated row into an insert row and a delete row.  Static helper method `MergeProcessorUtilities.createMergedDeleteAndInsertPages(Page inputPage, int dataColumnCount)` is available to connector merge sinks to split the page produced by `DeleteAndInsertMergeProcessor`, containing only inserts and deletes into a page of inserts and a page of deletes.

> * what is table handle lifecycle for MERGE. for example, how MERGE interacts with partition, file and file chunk pruning

I don't understand the question.  Could you be more specific?

Reply:
seems i hijacked some other thread, sorry

> > how should a connector construct rowIds if it needs to create deletion delta files for the sake of updates (e.g. a separate deletion file for an input file which would mark all the rows that got updated)
> 
> That's the point of the (poorly named) DeleteAndInsertMergeProcessor - - given a page that contains updated rows, it splits each updated row into an insert row and a delete row.

My question was a bit higher level that this. I should be clearer upfront.

I am primarily concerned about routing. 
If `number of input files > worker nodes`, after redistribution a node will have deletion data for many input files.
If it needs to create deletion deltas on per-input-file basis, it would need to sort that whole data, so cannot operate in streaming fashion.

This may be unfeasible especially if one of the input files is big.

BTW can redistribution columns reference portion of rowId? 

> > what is table handle lifecycle for MERGE. for example, how MERGE interacts with partition, file and file chunk pruning
>
> I don't understand the question. Could you be more specific?

for example --
If a MERGE updates a single row (source table is one row, or Merge's join condition is very pointed), what will be the amount of data scanned by the connector. In particular, will applyFilter() be called to avoid full table scan?

Reply:
The MERGE is built out of a Searched CASE on top of a RIGHT JOIN.  All the usual JOIN optimizations will apply.  For the Hive connector, using the `DeleteAndInsertRowParadigm`, the source is broadcast to the nodes handling the target, so the result will be full scan of the source but not the target.